### PR TITLE
feat(hogql): cohort left join conjoined

### PR DESCRIPTION
**Pull request by Gilbert09**
## Problem
- We were having an issue with big clients using multiple cohort breakdowns. We'd run multiple sequential queries, one for each cohort breakdown due to the breakdown value not getting returned from the `subquery` cohort method. 
- Introduced here https://github.com/PostHog/posthog/pull/18028 with a bit more detail on why it was built this way

## Changes
- Adds a new cohort modifier, `leftjoin_conjoined`. Built very similar to the `leftjoin` modifier, but instead of having a `LEFT JOIN` for **each cohort breakdown**, we instead have a single `LEFT JOIN` for all cohorts
- This allows us to return the cohort id as the breakdown value in the trends query when the modifier is turned on
- The new cohort modifier logic gets ran **before** the resolver is ran in hogql land. This is because we reference the join in the trends AST and so the left join needs to be added before types get resolved. 
- This PR **does not include** turning on the modifier in any way. This is to come in a follow-up PR. Question: how do we want to conditionally turn on modifiers?!

## How did you test this code?
- Added unit tests / snapshots of the new cohort resolver

### CODE REVIEW

**Review by mariusandra — COMMENTED** [review]
jumping to a call, dumping thoughts

File: posthog/hogql/transforms/in_cohort.py [hunk 1]
```diff
@@ -18,6 +19,172 @@ def resolve_in_cohorts(
...
+                    SELECT person_id AS cohort_person_id, 1 AS matched, cohort_id
+                    FROM static_cohort_people
+                    WHERE {cohort_clause}
+                    UNION ALL
+                    SELECT person_id AS cohort_person_id, 1 AS matched, cohort_id
+                    FROM raw_cohort_people
+                    WHERE {cohort_clause}
+                    GROUP BY cohort_person_id, cohort_id, version
+                    HAVING sum(sign) > 0
+                """,
```
[the remark this answers is not in the document]
Comment (Gilbert09) [reply posthog/hogql/transforms/in_cohort.py:151]: Yeah, we can check if we're querying just static, just dynamic, or both for sure

**Review by mariusandra — APPROVED** [review]
I think this makes sense. I see now as well why multiple joins with an OR wouldn't have really worked. Thinking now, there probably would have been a way to get it working with array joins, but it'd then still make multiple queries to the same cohorts tables, and possibly cause other issues with other aggregations. So LGTM, but I'd still see if there's something to do about the typing int/float typing. Can probably be later as well.

Edit: also there are some failing tests.

File: posthog/hogql/transforms/test/__snapshots__/test_in_cohort.ambr [hunk 2]
```diff
@@ -1,4 +1,106 @@
...
+  SELECT events.event AS event 
+  FROM events LEFT JOIN (
+  SELECT person_static_cohort.person_id AS cohort_person_id, 1 AS matched, person_static_cohort.cohort_id AS cohort_id 
+  FROM person_static_cohort 
+  WHERE and(equals(person_static_cohort.team_id, 420), in(person_static_cohort.cohort_id, [1])) UNION ALL 
+  SELECT cohortpeople.person_id AS cohort_person_id, 1 AS matched, cohortpeople.cohort_id AS cohort_id 
+  FROM cohortpeople 
+  WHERE and(equals(cohortpeople.team_id, 420), in(cohortpeople.cohort_id, [1])) 
+  GROUP BY cohort_person_id, cohortpeople.cohort_id, cohortpeople.version 
+  HAVING ifNull(greater(sum(cohortpeople.sign), 0), 0)) AS __in_cohort ON equals(__in_cohort.cohort_person_id, events.person_id)
```
Comment (neilkakkar) [inline posthog/hogql/transforms/test/__snapshots__/test_in_cohort.ambr:15]: flyby: cohortpeople.sign is deprecated, shouldn't be used, this doesn't work properly here. Rather, we look at the version, if it's equal to the version on the cohort.

I see sign is still used in a few places for cohort calculations, but I think those places don't face this same issue 🙈 (for various reasons). In querying though, stick to version!

This means _somehow_ passing the current cohort version value to the query generator

Comment (neilkakkar) [reply posthog/hogql/transforms/test/__snapshots__/test_in_cohort.ambr:15]: basically, the problem here is some stale versions of the cohorts might have sum(sign) > 0 as well, which can then lead to extra persons being filtered, iirc.

Comment (Gilbert09) [reply posthog/hogql/transforms/test/__snapshots__/test_in_cohort.ambr:15]: @neilkakkar Cool, so, I need to grab `cohort.version` and filter `cohort_people` + removing the `sign` clause? 👌

Comment (neilkakkar) [reply posthog/hogql/transforms/test/__snapshots__/test_in_cohort.ambr:12]: Yep, should work well!

This means you can get rid of the group by too, which should speed things up a lot, just copy this query:

https://github.com/PostHog/posthog/blob/master/posthog/models/cohort/sql.py#L75

Comment (Gilbert09) [reply posthog/hogql/transforms/test/__snapshots__/test_in_cohort.ambr:12]: Dreamy, have added this in, but need to remove the group by. Thanks!

**Pull request merged**